torch/distributed/_tensor/experimental/attention.py

```diff
@@ -0,0 +1,452 @@
+        *op_info.local_args,  # type: ignore[arg-type]
+        **op_info.local_kwargs,  # type: ignore[arg-type]
+    )
+
+    return DTensor._op_dispatcher.wrap(local_results, output_sharding.output_spec)
+
+
+def _scaled_dot_product_flash_attention_backward(
+    mesh: DeviceMesh,
+    grad_out: torch.Tensor,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    out: torch.Tensor,
+    softmax_lse: torch.Tensor,
+    cum_seq_q: torch.Tensor,
+    cum_seq_k: torch.Tensor,
+    max_q: int,
+    max_k: int,
+    dropout_p: float,
+    is_causal: bool,
+    philox_seed: torch.Tensor,
+    philox_offset: torch.Tensor,
+    *,
+    scale: Optional[float] = None,
+) -> Tuple[torch.Tensor, ...]:
+    pg = mesh.get_group()
+    assert isinstance(pg, dist.ProcessGroup), "must be single dimension"
+    rank = dist.get_rank(pg)
+    size = dist.get_world_size(pg)
+
+    # rank 0 sends to rank 1, rank 1 sends to rank 2, ..., rank n-1 sends to rank 0
+    right_dsts = list(range(1, size)) + [0]
+
+    next_kv = None
+
+    out_grad_queries = []
+    out_grad_keys = []
+    out_grad_values = []
+
+    for i in range(size):
+        # overlap communication with compute
+        if next_kv is not None:
+            next_kv = ft_c.wait_tensor(next_kv)
+            key = next_kv[: key.numel()].reshape(key.shape)
+            value = next_kv[key.numel() :].reshape(value.shape)
+
+        if i < (size - 1):
+            next_kv = torch.cat([key.flatten(), value.flatten()])
+            next_kv = ft_c.permute_tensor(next_kv, right_dsts, pg)
+
+        is_causal_behavior = _is_causal_behavior(
+            rank=rank, world_size=size, i=i, is_causal=is_causal
+        )
+
+        if is_causal_behavior != _CausalBehavior.SKIP:
+            # we rerun the forwards pass since we don't have a good way to save the
+            # output/logsumexp
+            (
+                output,
+                logsumexp,
+                cum_seq_q,
+                cum_seq_k,
+                max_q,
+                max_k,
+                philox_seed,
+                philox_offset,
+                _,
+            ) = torch.ops.aten._scaled_dot_product_flash_attention(
+                query,
+                key,
+                value,
+                dropout_p=dropout_p,
+                is_causal=is_causal_behavior.value,
+                scale=scale,
+            )
+
+            softmax_lse_corrected = torch.exp(logsumexp - softmax_lse)
+
+            chunk_grad = grad_out * softmax_lse_corrected.conj().unsqueeze(-1).to(
+                grad_out.dtype
+            )
+
+            (
+                grad_query,
+                grad_key,
+                grad_value,
+            ) = torch.ops.aten._scaled_dot_product_flash_attention_backward(
+                grad_out=chunk_grad,
+                query=query,
+                key=key,
+                value=value,
+                out=output,
+                logsumexp=logsumexp,
```

Comment:
I need to some math but do you know if you need to use the corrected LSE?

Reply:
We need the non-corrected LSE (from the forwards pass), the corrected LSE is used to calculate chunk_grad. I've confirmed this has the correct behavior by using the chunked autogenerated backwards pass